aws/table_aws_pricing_product.go

```diff
@@ -0,0 +1,274 @@
+	// Create Session
+	svc, err := PricingServiceClient(ctx, d)
+	if err != nil {
+		plugin.Logger(ctx).Error("aws_pricing_product.listPricingProduct", "connection_error", err)
+		return nil, err
+	}
+	if svc == nil {
+		// Unsupported region, return no data
+		return nil, nil
+	}
+
+	// Limiting the results
+	maxItems := int32(100)
+	if d.QueryContext.Limit != nil {
+		limit := int32(*d.QueryContext.Limit)
+		if limit < maxItems {
+			if limit < 20 {
+				maxItems = 20
+			} else {
+				maxItems = limit
+			}
+		}
+	}
+
+	equalQual := d.KeyColumnQuals
+	input := pricing.GetProductsInput{
+		ServiceCode:   aws.String(equalQual["service_code"].GetStringValue()),
+		FormatVersion: aws.String("aws_v1"),
+		MaxResults:    maxItems,
+	}
+
+	type OfferOutput struct {
+		_              struct{} `type:"structure"`
+		PriceDimension *PriceDimension
+		Term           string
+		EffectiveDate  *time.Time
+		OfferTermCode  *string
+		TermAttributes *TermAttributes
+	}
+
+	type PriceOutput struct {
+		_               struct{} `type:"structure"`
+		Product         *Product
+		ServiceCode     *string
+		Offer           *OfferOutput
+		Version         *string
+		PublicationDate *time.Time
+	}
+
+	filters := buildPricingFilter(d.Quals)
+	if len(filters) > 0 {
+		input.Filters = filters
+	}
+
+	paginator := pricing.NewGetProductsPaginator(svc, &input, func(o *pricing.GetProductsPaginatorOptions) {
+		o.Limit = maxItems
+		o.StopOnDuplicateToken = true
+	})
+
+	for paginator.HasMorePages() {
+		output, err := paginator.NextPage(ctx)
+		if err != nil {
+			plugin.Logger(ctx).Error("aws_pricing_product.listPricingProduct", "api_error", err)
+			return nil, err
+		}
+
+		for _, priceList := range output.PriceList {
+			var priceListObject PriceList
+
+			err = json.Unmarshal([]byte(priceList), &priceListObject)
+			if err != nil {
+				plugin.Logger(ctx).Error("listPricingProduct unmarshal", "err", err)
+				return nil, err
+			}
+
+			for term, termDetail := range priceListObject.Terms {
+				for _, offer := range *termDetail {
+					for _, priceDimension := range offer.PriceDimensions {
+						d.StreamListItem(ctx, PriceOutput{
+							Product:         priceListObject.Product,
+							ServiceCode:     priceListObject.ServiceCode,
+							Version:         priceListObject.Version,
+							PublicationDate: priceListObject.PublicationDate,
+							Offer:           &OfferOutput{PriceDimension: priceDimension, Term: term, EffectiveDate: offer.EffectiveDate, OfferTermCode: offer.OfferTermCode, TermAttributes: offer.TermAttributes},
+						})
+					}
+				}
+			}
+
+			// Context may get cancelled due to manual cancellation or if the limit has been reached
+			if d.QueryStatus.RowsRemaining(ctx) == 0 {
+				return nil, nil
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func extractPrice(_ context.Context, d *transform.TransformData) (interface{}, error) {
+	priceMap := d.Value.(map[string]*string)
+
+	for _, units := range priceMap {
+		price, err := strconv.ParseFloat(*units, 32)
```

Comment:
What are some examples of prices we get back from the API? Based on the suggested change to the column type, I think we wouldn't need to cast anymore?

Reply:
Hi @cbruno10 and thanks again for your high qualified review :-)

See bellow an example of price get back from the API : 

``` json
"VJKEFXHAX2JKR9HH.JRTCKXETXF.6YS6EN2CT7": {
            "unit": "Hrs",
            "endRange": "Inf",
            "description": "$0.404 per Unused Reservation Linux c5.2xlarge Instance Hour",
            "appliesTo": [],
            "rateCode": "VJKEFXHAX2JKR9HH.JRTCKXETXF.6YS6EN2CT7",
            "beginRange": "0",
            "pricePerUnit": {
              "USD": "0.4040000000"
            }
```

So, pricePerUnit is a map, with only one key:value row.
This is why i return the first record in my for loop : 

``` go
func extractCurrency(_ context.Context, d *transform.TransformData) (interface{}, error) {
	priceMap := d.Value.(map[string]*string)

	for currency := range priceMap {
		return currency, nil
	}
	return nil, nil
}
```

Feel free to ask me any change if you have a better idea

Reply:
Does it always just return `USD`, or does it sometimes return the local currency, e.g., `CAD` as a key? Do you know if there are any cases where it could return multiple price per unit currencies?

Reply:
We have always paid aws with `USD` :laughing:

More seriously, you can choose you currency in your aws account, but it seems everything is not converted magicaly.
See [here](https://aws.amazon.com/premiumsupport/knowledge-center/non-us-currency/?nc1=h_ls)

> When using a non-USD currency, consider the following:
> All preferred currency amounts shown in the AWS Billing and Cost Management console are estimated in USD.

So, it's not planned to have several currencies in pricing API apparently. I can hardcode `USD` key in the map if you want.

Reply:
I think it's ok to pull the first currency from the map if we only expect one item to be returned in the map.